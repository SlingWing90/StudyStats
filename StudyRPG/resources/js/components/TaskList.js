    import axios from 'axios'
    import React, { Component } from 'react'
    import PropTypes from 'prop-types'; 

    class TaskList extends Component {
      constructor (props) {
          super(props);
          
          this.state = {
              timeLeft: []
          }
          
          this.handler = this.handler.bind(this);
          this.calculateCountdown = this.calculateCountdown.bind(this);
          this.formatCountdown = this.formatCountdown.bind(this);
          this.timePercentage = this.timePercentage.bind(this);
      }
      
        
      calculateCountdown(endDate){
          let diff = (Date.parse(new Date(endDate)) - Date.parse(new Date())) / 1000;
          
          if(diff <= 0){
              return false;
          }
          
          const timeLeft = {
              years: 0,
              days: 0,
              hours: 0,
              minutes: 0,
              seconds: 0
          }
          
          if(diff >= (365 * 86400)){
              timeLeft.years = Math.floor(diff / (365 * 86400));
              diff -= timeLeft.years * 365 * 86400;
          }
          
          if(diff >= 86400){
              timeLeft.days = Math.floor(diff / (86400));
              diff -= timeLeft.days * 86400;
          }
          
          if(diff >= 3600){
              timeLeft.hours = Math.floor(diff / (3600));
              diff -= timeLeft.hours * 3600;
          }
          
          if(diff >= 60){
              timeLeft.minutes = Math.floor(diff / (60));
              diff -= timeLeft.minutes * 60;
          }
          
          timeLeft.seconds = diff;
          
          // this.setState({timeLeft: timeLeft});
          return timeLeft;
          
          //console.log(days);
      }
        
      formatCountdown(end){
          const timeLeft = this.calculateCountdown(end);
          
          //console.log(timeLeft);
          let timeLeftString = "";
          if(timeLeft != false){
              timeLeftString = "Noch ";
          }else{
              timeLeftString = "Abgelaufen!";
          }
          
          if(timeLeft.years > 0){
              timeLeftString += timeLeft.years+" Jahre";
          }
          if(timeLeft.days > 0){
              timeLeftString += " "+timeLeft.days+" Tage";
          }
          if(timeLeft.hours > 0){
              timeLeftString += " "+timeLeft.hours+" Stunden";
          }          
          if(timeLeft.minutes > 0){
              timeLeftString += " "+timeLeft.minutes+" Minuten";
          } 
          if(timeLeft.seconds > 0){
              timeLeftString += " "+timeLeft.seconds+" Sekunden";
          } 
          
          
          return timeLeftString; //timeLeft.days+" Tage "+timeLeft.hours
      }
        
      timePercentage(start, end){
          //let diff = (Date.parse(new Date(endDate)) - Date.parse(new Date())) / 1000;
          let endDate = Date.parse(new Date(end));
          let startDate = Date.parse(new Date(start));
          let actDate = Date.parse(new Date());
          
          let left = endDate - actDate;
          let total = endDate - startDate;
          
          //const percentage = 100 - ((left / total) * 100);
          
          const percentage = 100 - (100 / total) * left;
          
          if(percentage >= 100){
              return 100;
          }
          
          return percentage;
      }
        
      handler(e){
          this.props.onDoneClick(e);
      }
        
      render () {
          
        const {tasks} = this.props
          
        return (
          <div className="col-md-6">
            <h2 className="text-center">Aufgaben</h2>
            <ul className="list-group task-list">   
                {tasks.map((item, key) =>
                    <li className="list-group-item" key={key}>
                        <strong>{item.name}</strong>
                        <div>
                            {item.description}
                        </div>
                        <div>
                            <div className="progressbar">
                                <div className="bar" style={{width: this.timePercentage(item.start, item.end)+"%"}}>
                                </div>
                            </div>
                            <small className="duration-text">{this.formatCountdown(item.end)}</small>
                        </div>
                        
                        <button type="button" onClick={() => this.handler(key)} className="btn btn-primary margin-top-10">Erledigt</button>
                    </li>
                )}
            </ul>
        </div>
        )
      }
    }
    
    TaskList.propTypes = {
      tasks: PropTypes.array,
      onDoneClick: PropTypes.func
    };

    export default TaskList