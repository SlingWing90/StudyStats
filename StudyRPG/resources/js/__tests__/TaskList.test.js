import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import TaskList from '../components/TaskList';
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({adapter: new Adapter()});

describe('<TaskList />', () => {
    it('renders <TaskList /> component with empty props without crashing', () => {
        const clickFn = jest.fn();
        const component = shallow(<TaskList tasks={[]} onDoneClick={clickFn} />);  
        
        expect(component.find('.list-group-item')).toHaveLength(0);   
        expect(component).toMatchSnapshot();
    });
    
    it('renders <TaskList /> component with two tasks', () => {
        const clickFn = jest.fn();
        
        // 2 Tasks
        const test_tasks = [{"id":17,"subject_id":1,"name":"Sint debitis nihil","description":"Corporis modi","start":"2018-06-05 00:00:00","end":"2018-08-21 00:00:00","finished":"2019-06-18 00:00:00","done":0,"created_at":"2019-05-14 08:17:21","updated_at":"2019-05-14 08:17:21"},{"id":2,"subject_id":4,"name":"Harum deserunt accusantium","description":"Et quibusdam nemo","start":"2018-07-24 00:00:00","end":"2018-09-25 00:00:00","finished":"2019-04-09 00:00:00","done":0,"created_at":"2019-05-14 08:17:21","updated_at":"2019-05-14 08:17:21"}];
        
        const component = shallow(<TaskList tasks={test_tasks} onDoneClick={clickFn} />);
        
        // could be li or .list-group-item - li better?
        expect(component.find('.list-group-item')).toHaveLength(2);
    });
});