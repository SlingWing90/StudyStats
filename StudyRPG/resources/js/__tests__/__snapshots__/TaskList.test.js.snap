// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskList /> renders <TaskList /> component with empty props without crashing 1`] = `ShallowWrapper {}`;
